### PR TITLE
FIX-#6372: precompute dtypes for 'sum' operation

### LINKED ISSUE #6372: Precompute dtypes for `sum` operation

### PR DESCRIPTION
**Pull request by anmyachev**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6372 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/dataframe/algebra/tree_reduce.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ class TreeReduce(Operator):
     """Builder class for TreeReduce operator."""
 
     @classmethod
-    def register(cls, map_function, reduce_function=None, axis=None):
+    def register(cls, map_function, reduce_function=None, axis=None, dtypes=None):
```
Comment (vnlitvinov) [inline modin/core/dataframe/algebra/tree_reduce.py:23]: Maybe name it `compute_dtypes` to clarify it's a callable?

Comment (anmyachev) [reply modin/core/dataframe/algebra/tree_reduce.py]: Agree, changed

File: modin/core/dataframe/algebra/tree_reduce.py [hunk 2]
```diff
@@ -45,11 +49,17 @@ def register(cls, map_function, reduce_function=None, axis=None):
...
+            new_dtypes = None
+            if compute_dtypes and query_compiler._modin_frame.has_materialized_dtypes:
+                new_dtypes = str(compute_dtypes(query_compiler.dtypes, *args, **kwargs))
+
             return query_compiler.__constructor__(
                 query_compiler._modin_frame.tree_reduce(
                     cls.validate_axis(_axis),
                     lambda x: map_function(x, *args, **kwargs),
                     lambda y: reduce_function(y, *args, **kwargs),
+                    dtypes=new_dtypes,
```
Comment (vnlitvinov) [inline modin/core/dataframe/algebra/tree_reduce.py:62]: Don't you have to rename this argument as well?

Comment (anmyachev) [reply modin/core/dataframe/algebra/tree_reduce.py:62]: The type of this variable does not change, so I did not change its name.

**Review by vnlitvinov — APPROVED** [review]
LGTM!

**Pull request merged**